File: app/src/main/java/com/pr0gramm/app/util/Holder.java
package com.pr0gramm.app.util;

import com.google.common.base.Optional;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.SettableFuture;

import rx.Observable;
import rx.Single;
import rx.subjects.BehaviorSubject;

/**
 * Holds an {@link rx.Single} value and provides access
 * to the result.
 */
public class Holder<T> {
    private final BehaviorSubject<T> subject = BehaviorSubject.create();
    private SettableFuture<T> future = SettableFuture.create();

    private Holder(Single<T> single) {
        single.subscribe(
                value -> {
                    future.set(value);
                    subject.onNext(value);
                },
                error -> {
                    future.setException(error);
                    subject.onError(error);
                });
    }

    public Single<T> asSingle() {
        return subject.take(1).toSingle();
    }

    public Observable<T> asObservable() {
        return subject.take(1);
    }

    public Optional<T> asOptional() {
        return future.isDone() ? Optional.of(value()) : Optional.absent();
    }

    public T value() {
        return Futures.getUnchecked(future);
    }

    public static <T> Holder<T> ofSingle(Single<T> single) {
        return new Holder<>(single);
    }

    public static <T> Holder<T> ofObservable(Observable<T> observable) {
        return ofSingle(observable.toSingle());
    }
}

File: app/src/main/java/com/pr0gramm/app/util/Holder.kt
package com.pr0gramm.app.util

import com.pr0gramm.app.ui.base.AsyncScope
import kotlinx.coroutines.async

/**
 * Holds a deferred value and provides access to the result.
 */
class Holder<T>(provider: suspend () -> T) {
    private val deferred = AsyncScope.async { provider() }

    /**
     * Returns the current value or returns null, if the value is
     * not yet available.
     */
    val valueOrNull: T?
        get() {
            if (deferred.isActive || deferred.isCancelled)
                return null

            return runCatching { deferred.getCompleted() }.getOrNull()
        }

    suspend fun get(): T = deferred.await()
}
